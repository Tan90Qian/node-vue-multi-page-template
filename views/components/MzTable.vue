<template>
  <div class="mz-table">
    <div class="mz-table-header">
      <table :border="bordered ? undefined : 0" :cellspacing="bordered ? undefined : 0">
        <thead>
          <th
            v-for="(column, index) in columns"
            :key="column.key || column.dataIndex || index"
            :width="column.width"
          >
            {{ column.title }}
          </th>
        </thead>
      </table>
    </div>
    <div class="mz-table-body">
      <table
        v-if="dataSource.length"
        :border="bordered ? undefined : 0"
        :cellspacing="bordered ? undefined : 0"
      >
        <tbody>
          <tr
            v-for="(item, index) in dataSource"
            :key="rowKey ? item[rowKey] : index"
            :class="rowClassName"
            :style="{ cursor: typeof onRowClick === 'function' ? 'pointer' : 'default' }"
            @click="typeof onRowClick === 'function' ? onRowClick(item, index) : undefined;"
          >
            <td
              v-for="(column, index) in columns"
              :key="column.key || column.dataIndex || index"
              :width="column.width"
            >
              {{
                typeof column.render === 'function'
                  ? column.render(item[column.dataIndex], item, index)
                  : item[column.dataIndex]
              }}
            </td>
          </tr>
        </tbody>
      </table>
      <div v-else class="no-data">暂无数据</div>
    </div>
  </div>
</template>
<script>
export default {
  props: {
    columns: {
      type: Array,
      required: true
    },
    dataSource: {
      type: Array,
      default: function () {
        return [];
      }
    },
    rowKey: {
      type: [String, Number]
    },
    bordered: {
      type: Boolean,
      default: false
    },
    rowClassName: {
      type: String
    },
    onRowClick: {
      type: Function
    }
  }
};
</script>

<style lang="scss">
.mz-table {
  table {
    width: 100%;
    color: #212121;
    line-height: 58px;
    text-align: center;
    thead {
      background: #d7dbe9;
      font-weight: bold;
      font-size: 18px;
    }
    tbody {
      background: #fff;
      font-size: 16px;
      tr:nth-of-type(odd) {
        background: #f5f5f8;
      }
    }
  }
  .no-data {
    color: #212121;
    line-height: 60px;
    text-align: center;
  }
}
</style>
