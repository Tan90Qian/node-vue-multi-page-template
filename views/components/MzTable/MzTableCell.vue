<script>
import { isValidElement } from '../_util/prop-util.js';

function isInvalidRenderCellText(text) {
  return (
    text && !isValidElement(text) && Object.prototype.toString.call(text) === '[object Object]'
  );
}

export default {
  name: 'MzTableCell',
  props: {
    record: Object,
    column: Object,
    index: Number
  },
  methods: {
    handleClick(e) {
      const {
        record,
        column: { onCellClick }
      } = this;
      if (onCellClick) {
        onCellClick(record, e);
      }
    }
  },
  render() {
    const { column, record, index } = this;
    const { dataIndex, customRender, className = '' } = column;
    let text;
    let colSpan;
    let rowSpan;
    const tdProps = {
      attrs: {},
      class: className || column.class,
      on: {
        click: this.handleClick
      }
    };
    if (typeof dataIndex === 'string' || typeof dataIndex === 'number') {
      text = record[dataIndex];
    } else if (!dataIndex || dataIndex.length === 0) {
      text = record;
    }
    if (customRender) {
      text = customRender.call(this, text, record, index);
      if (isInvalidRenderCellText(text)) {
        tdProps.attrs = text.attrs;
        colSpan = tdProps.attrs.colSpan;
        rowSpan = tdProps.attrs.rowSpan;
        text = text.children;
      }
    }
    if (isInvalidRenderCellText(text)) {
      text = null;
    }
    if (rowSpan === 0 || colSpan === 0) {
      return null;
    }
    if (column.align) {
      tdProps.style = { textAlign: column.align };
    }
    if (column.width) {
      tdProps.attrs.width = column.width;
    }
    return <td {...tdProps}>{text}</td>;
  }
};
</script>
