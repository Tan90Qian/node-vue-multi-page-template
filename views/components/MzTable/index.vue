<template>
  <div class="mz-table">
    <table :border="bordered ? 1 : 0" :cellspacing="bordered ? undefined : 0">
      <thead>
        <th
          v-for="(column, index) in columns"
          :key="column.key || column.dataIndex || index"
          :width="column.width"
        >
          {{ column.title }}
        </th>
      </thead>
      <tbody>
        <tr
          v-for="(record, rowIndex) in dataSource"
          :key="rowKey ? record[rowKey] : rowIndex"
          :class="
            typeof rowClassName === 'function' ? rowClassName(record, rowIndex) : rowClassName
          "
          :style="{ cursor: typeof onRowClick === 'function' ? 'pointer' : 'default' }"
          @click="typeof onRowClick === 'function' ? onRowClick(record, rowIndex) : undefined;"
        >
          <mz-table-cell
            v-for="(column, colIndex) in columns"
            :key="column.key || column.dataIndex || colIndex"
            :column="column"
            :record="record"
            :index="rowIndex"
          ></mz-table-cell>
        </tr>
      </tbody>
    </table>
    <div v-if="!dataSource || !dataSource.length" class="no-data">暂无数据</div>
  </div>
</template>
<script>
import MzTableCell from './MzTableCell';

export default {
  name: 'MzTable',
  components: {
    MzTableCell
  },
  props: {
    columns: {
      type: Array,
      required: true
    },
    dataSource: {
      type: Array,
      default: function () {
        return [];
      }
    },
    rowKey: [String, Number],
    bordered: {
      type: Boolean,
      default: false
    },
    rowClassName: [String, Function],
    onRowClick: Function
  }
};
</script>

<style lang="scss">
.mz-table {
  table {
    width: 100%;
    color: #212121;
    line-height: 58px;
    text-align: center;
    thead {
      background: #d7dbe9;
      font-weight: bold;
      font-size: 18px;
    }
    tbody {
      background: #fff;
      font-size: 16px;
      tr:nth-of-type(odd) {
        background: #f5f5f8;
      }
    }
  }
  .no-data {
    color: #212121;
    line-height: 60px;
    text-align: center;
  }
}
</style>
