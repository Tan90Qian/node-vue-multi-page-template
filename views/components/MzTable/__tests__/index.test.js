import { mount } from '@vue/test-utils';

import MzTable from '../index';

describe('MzTable tests', () => {
  it('should render .mz-table', () => {
    const wrapper = mount(MzTable, {
      propsData: {
        columns: []
      }
    });
    const mzTable = wrapper.find('.mz-table');

    expect(mzTable.exists()).toBe(true);
  });

  it('should render .no-data when dataSource was absent', () => {
    const wrapper = mount(MzTable, {
      propsData: {
        columns: []
      }
    });
    const noData = wrapper.find('.no-data');
    expect(noData.exists()).toBe(true);
  });

  it('should show border and cellspacing when bordered props was true', () => {
    const wrapper = mount(MzTable, {
      propsData: {
        columns: [],
        dataSource: [{}],
        bordered: true
      }
    });
    const table = wrapper.find('table');
    expect(table.attributes('border')).toBe('1');
    expect(table.attributes('cellspacing')).toBeFalsy();
  });

  it('should hide border and cellspacing when bordered props was false or absent', () => {
    const wrapper = mount(MzTable, {
      propsData: {
        columns: [],
        dataSource: [{}]
      }
    });
    const table = wrapper.find('table');
    expect(table.attributes('border')).toBe('0');
    expect(table.attributes('cellspacing')).toBe('0');
  });

  it('should have correct number of th', () => {
    const wrapper = mount(MzTable, {
      propsData: {
        columns: [{ dataIndex: 'a' }, { dataIndex: 'b' }, { dataIndex: 'c' }]
      }
    });
    const thArray = wrapper.findAll('th');

    expect(thArray.length).toBe(3);
  });

  it('should have correct number of tr', () => {
    const wrapper = mount(MzTable, {
      propsData: {
        columns: [],
        dataSource: [{ a: 1 }, { a: 2 }, { a: 3 }]
      }
    });
    const trArray = wrapper.findAll('tr');

    expect(trArray.length).toBe(3);
  });

  it("column's width should act on td's width", () => {
    const wrapper = mount(MzTable, {
      propsData: {
        columns: [
          {
            dataIndex: 'a',
            title: 'a',
            width: 200
          },
          {
            dataIndex: 'b',
            title: 'b'
          }
        ],
        dataSource: [{ a: 1, b: 1 }]
      }
    });
    const tdArray = wrapper.findAll('td');

    expect(tdArray.at(0).attributes('width')).toBe('200');
    expect(tdArray.at(1).attributes('width')).toBeFalsy();
  });
});
