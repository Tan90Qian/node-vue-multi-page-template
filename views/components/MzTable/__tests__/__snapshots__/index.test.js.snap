// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MzTable tests renders correctly 1`] = `
<div
  class="mz-table"
>
  <table
    border="0"
    cellspacing="0"
  >
    <thead />
     
    <tbody />
  </table>
   
  <div
    class="no-data"
  >
    暂无数据
  </div>
</div>
`;
